### PR TITLE
Add AddingTrailingDataSubscriber to allow users to send additional data t…

### PR DESCRIPTION
**Pull request by zoewangg**
…o the downstream subscriber

## Motivation and Context
There are use-cases where we want to send trailing data on top of what the publisher publishes. For example, in async signing code path, we always send a trailing empty frame.

## Modifications
- Add AddingTrailingDataSubscriber to allow users to send trailling to the downstream subscriber
- Expose `SdkPublisher#addTrailingData` 

## Testing
Added TCK tests and unit tests.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [x] My change requires a change to the Javadoc documentation
- [x] I have updated the Javadoc documentation accordingly
- [ ] I have added tests to cover my changes
- [x] All new and existing tests passed
- [ ] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: utils/src/main/java/software/amazon/awssdk/utils/async/AdditionalDataSubscriber.java [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    }
+
+    @Override
+    public void onSubscribe(Subscription subscription) {
+
+        if (upstreamSubscription != null) {
+            log.warn(() -> "Received duplicate subscription, cancelling the duplicate.", new IllegalStateException());
+            subscription.cancel();
+            return;
+        }
```
Comment (cenedhryn) [inline utils/src/main/java/software/amazon/awssdk/utils/async/AdditionalDataSubscriber.java:73]: Can we test this?

Comment (zoewangg) [reply utils/src/main/java/software/amazon/awssdk/utils/async/AdditionalDataSubscriber.java]: This is covered in TCK tests :) `required_spec205_mustCallSubscriptionCancelIfItAlreadyHasAnSubscriptionAndReceivesAnotherOnSubscribeSignal`

File: utils/src/main/java/software/amazon/awssdk/utils/async/AddingTrailingDataSubscriber.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+import software.amazon.awssdk.annotations.SdkProtectedApi;
+import software.amazon.awssdk.utils.Logger;
+import software.amazon.awssdk.utils.Validate;
+
+/**
+ * Allows to send trailing data before invoking onComplete on the downstream subscriber.
+ * If the trailingDataSupplier returns null, this class will invoke onComplete directly
+ */
+@SdkProtectedApi
+public class AddingTrailingDataSubscriber<T> extends DelegatingSubscriber<T, T> {
```
Comment (dagnir) [inline utils/src/main/java/software/amazon/awssdk/utils/async/AddingTrailingDataSubscriber.java:32]: I think this is great. Would it be much more work to to generalize this to work for N number additional data?

File: utils/src/main/java/software/amazon/awssdk/utils/async/AddingTrailingDataSubscriber.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        downstreamDemand.decrementAndGet();
+        subscriber.onNext(t);
+    }
+
+    @Override
+    public void onComplete() {
+        onCompleteCalledByUpstream = true;
+
+        trailingData = trailingDataSupplier.get();
+        if (trailingData == null || downstreamDemand.get() > 0) {
```
Comment (L-Applin) [inline utils/src/main/java/software/amazon/awssdk/utils/async/AddingTrailingDataSubscriber.java:122]: We `get()` the trailing data from the supplier, but send it only if the supplier returns `null`? Is this really what we want?

File: utils/src/main/java/software/amazon/awssdk/utils/async/AddingTrailingDataSubscriber.java (hunk 2 above)
Comment (zoewangg) [reply utils/src/main/java/software/amazon/awssdk/utils/async/AddingTrailingDataSubscriber.java:32]: Good idea! Updated

File: utils/src/main/java/software/amazon/awssdk/utils/async/AddingTrailingDataSubscriber.java (hunk 3 above)
Comment (zoewangg) [reply utils/src/main/java/software/amazon/awssdk/utils/async/AddingTrailingDataSubscriber.java]: This code has been updated, but to answer your question, the original code allows it to trigger onComplete immediately if trailingData is null

**Review by cenedhryn — APPROVED** [review]

**Review by L-Applin — APPROVED** [review]

**Review by dagnir — APPROVED** [review]

**Pull request merged**